### PR TITLE
[MPS] Vectorized contiguous cat fast path for any dimension

### LINKED ISSUE #188198: [MPS] cat op dropped performance in torch==2.10 release
### 🐛 Describe the bug

Cat op perf regressed in #165373 (#166556 regained the perf a little bit), landing in torch==2.10.0. The change moved the op from going through MPSGraph to use metal kernels, and added a slightly faster path through the blit encoder for contiguous cases. Unfortunately the blit encoder seems not to be the best performance for this application as it seems to saturate at a much lower throughput than the compute encoder approach. Measurements from 2.9.1 and 2.10.0 (the nightly still has the same issue as 2.10.0) on an M2 Max using the benchmark script attached below. The GB/s column is the important one for this issue:

Before:
```
     ### torch 2.9.1 (MPSGraph cat - pre-Metal) ###
     torch 2.9.1    ~268MB/case, GB/s = read inputs + write output
             tier                     shapes  dim  gcd    GB/s
      uint4 (16B)                  8192+8192    1   16     301
      uint2 (8B)                   8194+8194    1    8     298
      uint  (4B)                   8191+8191    1    4     294
      ushort (2B)                16383+16383    1    2     198
      byte  (1B)                 16383+16384    1    1     225
      outer dim0                   8192+8192    0   16     305
```

After:
```
     ### torch 2.10.0 (Metal+blit cat - first regressed release) ###
     torch 2.10.0    ~268MB/case, GB/s = read inputs + write output
             tier                     shapes  dim  gcd    GB/s
      uint4 (16B)                  8192+8192    1   16     160
      uint2 (8B)                   8194+8194    1    8     160
      uint  (4B)                   8191+8191    1    4     159
      ushort (2B)                16383+16383    1    2      82
      byte  (1B)                 16383+16384    1    1      82
      outer dim0                   8192+8192    0   16     128
```

[cat_sweep.py](https://github.com/user-attachments/files/29357988/cat_sweep.py)

Opening a PR to move the metal kernel implementation to go through an alignment tiered compute kernel for contiguous case so we'll get the perf back closer to the hardware bandwidth limit.

### Versions

Tested on 2.9.1 (before regression), 2.10.0 (after regression) and nightly (that still has the issue). Verified on macOS27 (beta) and macos26.5.

cc @kulinseth @malfet @DenisVieriu97 @aditvenk @Isalia20

### PR DESCRIPTION
**Pull request by jhavukainen**
Fixes https://github.com/pytorch/pytorch/issues/188198

Note: This is staged on top of `dev/joona/copy_bytes_aligned` since there we'll use the refactored parts from it. So we need to land https://github.com/pytorch/pytorch/pull/188613. Unfortunately I worked on these in a backward order so the PR is not properly gh stacked.

Routes contiguous, same-dtype torch.cat through a vectorized inner_contiguous_scatter compute kernel that copies each input as raw bytes into its strided slice of the output (widest-aligned vector stores via copy_bytes_aligned). This replaces the general elementwise cat kernel for the common case and extends the fast path to any concat dimension not just dim 0.

The kernel, StridedBlockParams geometry and the inner_contiguous_scatter_mps host dispatcher live on the copy side (Copy.metal, Copy.h, Copy.mm) as reusable infrastructure for other kernels as well.

Measuring before and after for `cat`.

torch.cat MPS bandwidth, ~268MB/case, GB/s = read inputs + write output, best-of-50.
```
  ┌─────────────┬─────┬──────────────────┬─────────────────────┬─────────────────────────┐
  │    tier     │ gcd │ 2.9.1 (MPSGraph) │ 2.10.0 (Metal+blit) │ this PR (Metal+compute) │
  ├─────────────┼─────┼──────────────────┼─────────────────────┼─────────────────────────┤
  │ uint4 (16B) │  16 │              301 │                 160 │                     329 │
  ├─────────────┼─────┼──────────────────┼─────────────────────┼─────────────────────────┤
  │ uint2 (8B)  │   8 │              298 │                 160 │                     319 │
  ├─────────────┼─────┼──────────────────┼─────────────────────┼─────────────────────────┤
  │ uint (4B)   │   4 │              294 │                 159 │                     318 │
  ├─────────────┼─────┼──────────────────┼─────────────────────┼─────────────────────────┤
  │ ushort (2B) │   2 │              198 │                  82 │                     211 │
  ├─────────────┼─────┼──────────────────┼─────────────────────┼─────────────────────────┤
  │ byte (1B)   │   1 │              225 │                  82 │                     255 │
  ├─────────────┼─────┼──────────────────┼─────────────────────┼─────────────────────────┤
  │ outer dim0  │  16 │              305 │                 128 │                     320 │
  └─────────────┴─────┴──────────────────┴─────────────────────┴─────────────────────────┘
```

[cat_sweep.py](https://github.com/user-attachments/files/29358669/cat_sweep.py)

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @mcarilli @ptrblck @leslie-fang-intel @EikanWang @voznesenskym @penguinwu @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98 @xmfan @aditvenk

### CODE REVIEW

**Comment by jhavukainen** [comment]
Same measurements on an M5 Max to see the effect across hardware generations. Notably the blit encoder seems to catch-up on the newer hardware so the effect is not so pronounced which is probably why we didn't catch it initially when the op implementation moved. But the fix here is still an improvement across the board.

```
  +--------------+-------------+-----+-----+------------------+---------------------+-------------------------+
  | tier         | shapes      | dim | gcd | 2.9.1 (MPSGraph) | 2.10.0 (Metal+blit) | this fix (Metal+compute) |
  +--------------+-------------+-----+-----+------------------+---------------------+-------------------------+
  | uint4  (16B) | 8192+8192   |  1  |  16 |       437        |         320         |          455            |
  | uint2  (8B)  | 8194+8194   |  1  |   8 |       357        |         318         |          418            |
  | uint   (4B)  | 8191+8191   |  1  |   4 |       346        |         316         |          426            |
  | ushort (2B)  | 16383+16383 |  1  |   2 |       370        |         178         |          426            |
  | byte   (1B)  | 16383+16384 |  1  |   1 |       370        |         179         |          433            |
  | outer dim0   | 8192+8192   |  0  |  16 |       450        |         448         |          455            |
  +--------------+-------------+-----+-----+------------------+---------------------+-------------------------+
```

**Comment by malfet** [comment]
@claude review this and specifically compare to existing copy to Metal migration, that should be doing the same. Is there any code overlap that could be leveraged?

**Comment by jhavukainen** [comment]
@malfet This part sounds like the potential part where we can reduce duplication for me

> Those are hitting the same blit that regressed cat. So cat_copy_contiguous is really a general fast contiguous device-to-device memcpy that the same-dtype copy path could share

I'll check the copy metal implementation and see if we get the expected bandwidth there or if this should be lifted to the copy kernel. But yeah calling into the copy kernel for the relevant part and making sure the performance there is good sounds like the way to go.

**Comment by jhavukainen** [comment]
Turns out the blit encoder was accidentally load bearing by virtue of implicitly ending kernel coalescing. The PR https://github.com/pytorch/pytorch/pull/188308 will remove that duty from it so we can move away from the blit encoder here.

**Comment by jhavukainen** [comment]
To summarize: Based on the discussion and further investigation, it'll be better to move cat to call into copy as it'll achieve the same throughput as a bespoke kernel here does. It is the same operation underneath. This PR only addresses the truly contiguous case which should be the ~most impactful. Will need to think about a bit on whats the correct abstraction for the "almost-contiguous" cases since it should be a pattern many kernels share and should be generalized.

**Comment by jhavukainen** [comment]
Didn't pick up the PR adding the end coalescing calls in the last rebase. Let's try that again.

@pytorchbot rebase

**Comment by jhavukainen** [comment]
@claude please review again. I did a bunch of refactoring w.r.t. other copy specific PR to disentangle some of the shareable functionality so the previous review is out-dated.

**Review by malfet — APPROVED** [review]
